L58-python-async-api.md

```diff
@@ -0,0 +1,1708 @@
+Future, Coroutine, Event-loops, like `gevent`/`Twisted`/`Tornado`. In general,
+it is challenging to support multiple async mechanisms inter-operate together,
+and it has to be discussed case by case.
+
+For `Tornado`, they are using `asyncio` underneath their abstraction of
+asynchronous mechanisms. After v6.0.0, they dropped support for Python 2
+entirely. So, it should work fine with our new API.
+
+For `Twisted`, with some monkey patch code to connect its `Deferred`/`Future`
+object to `asyncio`'s API, it should work.
+
+For `gevent`, unfortunately, it works by monkey-patching Python APIs that
+including `threading` and various transport library, and the work is executed in
+`gevent` managed event loop. `gevent` and `asyncio` doesn't compatible with each
+other out of box.
+
+### Wrapping Async Stack To Provide Current API Publicly
+
+1) It would be quite confusing to mix functions that return `asyncio` coroutine
+   and normal Python functions;
+2) We don't want to imply that switching to use the new stack requires zero code
+   changes.
+3) Also, the current contract of API constraint ourselves from changing their
+   behavior after GA for two years.
+
+### Explicit Context vs. Implicit Context
+
+TBD
+
+## API Interfaces
+
+### Channel-Side
+
+```Python
+# grpc.aio.Channel
+class Channel:
+    """Affords RPC invocation via generic methods on client-side.
+
+    Channel objects implement the Async Context Manager type, although they need
+    not support being entered and exited multiple times.
+    """
+
+    def subscribe(self,
+                  callback: Callable[[ChannelConnectivity], None],
+                  try_to_connect: bool=False) -> None:
+        """Subscribe to this Channel's connectivity state machine.
+
+        A Channel may be in any of the states described by ChannelConnectivity.
+        This method allows application to monitor the state transitions.
+        The typical use case is to debug or gain better visibility into gRPC
+        runtime's state.
+
+        Args:
+          callback: A callable to be invoked with ChannelConnectivity argument.
+            ChannelConnectivity describes current state of the channel.
+            The callable will be invoked immediately upon subscription
+            and again for every change to ChannelConnectivity until it
+            is unsubscribed or this Channel object goes out of scope.
+          try_to_connect: A boolean indicating whether or not this Channel
+            should attempt to connect immediately. If set to False, gRPC
+            runtime decides when to connect.
+        """
+
+    def unsubscribe(self,
+                    callback: Callable[[ChannelConnectivity], None]) -> None:
+        """Unsubscribes a subscribed callback from this Channel's connectivity.
+
+        Args:
+          callback: A callable previously registered with this Channel from
+          having been passed to its "subscribe" method.
+        """
```

Comment:
Can it be replaced by the following API, please?
```python
def check_connectivity_state(self, try_to_connect: bool=False) -> ChannelConnectivity:
    pass

async def watch_connectivity_state(self,
                                   last_observed_state: ChannelConnectivity,
                                   timeout_seconds: float) -> ChannelConnectivity:
    pass
```

It aligns better with grpc-core API, and `subscribe`/`unsubscribe` API can be easily built on top of this primitives outside of base grpc package.

Reply:
The API you proposed is much neat than existing version.

The only concern I have is that `subscribe`/`unsubscribe` are not marked experimental. I think we need keep it. But another problem will be introduced, since the subscribe and the new API doesn't work well together. The underlying C-Core propagation of connectivity state can only be observed by one watch. If users is using both API, their application will have undefined behavior.

As a framework, we can document that don't use both API at the same time, and potentially perform check for users.

WDYT?

Reply:
Also, in the `watch_connectivity_state`, we need to return something when timeout, so I make it return `Optional[grpc.ChannelConnectivity]`.

Reply:
> The only concern I have is that subscribe/unsubscribe are not marked experimental. I think we need keep it.

We're designing **the new API in entirely different namespace** here (`grpc.aio.Channel`), so promises about the old API should not be a concern.

It's not breaking any users of the old API (as they have to make explicit code changes to move to async API), and it covers them by providing the same functionality via different means. It means that **old methods can simply be removed**.

_Of course the points would be different if this proposal discusses modifications of `grpc.Channel` API, but I have an impression that it's not the case or goal._

> The underlying C-Core propagation of connectivity state can only be observed by one watch.

Based on the documentation and implementation of `grpc_channel_watch_connectivity_state` it looks like it's safe to call it by multiple users. What makes you believe that it's not the case?

> Also, in the watch_connectivity_state, we need to return something when timeout, so I make it return Optional[grpc.ChannelConnectivity].

Yeah, I agree.

Reply:
The C-Core `grpc_channel_watch_connectivity_state` does allow multiple watcher. I was thinking about the implementation in Python wrapper, which only have one completion queue for the connectivity state propagation.

It shouldn't matter in Async IO ecosystem, since it will propagate through callback instead of `cq_next`.

I have written some application using the connectivity API, I would say in some cases the "watch" is more effective, and in some cases e.g. if you want some sort of global handler for certain connectivity state, the "subscribe" semantic is more handy.

For the sake of simplicity, probably we should remove the old API.

@gnossen WDTY about the old connectivity API?

Reply:
I think @euroelessar is right. The old API inverted core's pull API into a push API. This was particularly bad for the sync API because it spun up a thread outside of the [application author's control](https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_channel.py#L1011). You're always going to need another concurrent process (used in the pure concurrent computing sense, e.g. thread, fiber, coroutine, or OS process) to invert that model. This is perhaps not quite as bad in the asyncio case, since we're not preempting the application to run this poller, but it's still not ideal.

I'm leaning towards providing a pull model mirroring C-core's and letting application author's invert the model if they so choose. We can add an example of this in the examples directory.

As @euroelessar pointed out, we have no obligation to maintain "backwards compatibility" with the new set of APIs, so this change seems relatively innocuous to me.

In regards to the timeout, I'll also throw in that we could raise an exception. Either option is fine so long as all of your timeouts behave the same way across the asyncio API surface.

Reply:
As discussed offline, the old API will be removed, and we shall write an example to demonstrate how to implement the "subscribe" / "unsubscribe".